drivers/storage/portworx/status.go

```diff
@@ -79,20 +151,83 @@ func (p *portworx) UpdateStorageClusterStatus(
 		return fmt.Errorf("empty ClusterInspect response")
 	}
 
-	newClusterStatus := mapClusterStatus(pxCluster.Cluster.Status)
-	if cluster.Status.Phase != string(corev1.ClusterOnline) &&
-		newClusterStatus == corev1.ClusterOnline {
+	runtimeStatus := mapPortworxRuntimeStatus(pxCluster.Cluster.Status)
+	condition := util.GetStorageClusterCondition(cluster, constants.PortworxComponentName, corev1.ClusterConditionTypeRuntimeState)
+	if (condition == nil || condition.Status != corev1.ClusterConditionStatusOnline) &&
+		runtimeStatus == corev1.ClusterConditionStatusOnline {
 		msg := fmt.Sprintf("Storage cluster %v online", cluster.GetName())
 		p.normalEvent(cluster, util.ClusterOnlineReason, msg)
 	}
 
-	cluster.Status.Phase = string(newClusterStatus)
+	util.UpdateStorageClusterCondition(cluster, &corev1.ClusterCondition{
+		Source: constants.PortworxComponentName,
+		Type:   corev1.ClusterConditionTypeRuntimeState,
+		Status: runtimeStatus,
+	})
+
 	cluster.Status.ClusterName = pxCluster.Cluster.Name
 	cluster.Status.ClusterUID = pxCluster.Cluster.Id
 
 	return p.updateStorageNodes(cluster)
 }
 
+// getPortworxConditions returns a map of portworx conditions with type as keys
+func getPortworxConditions(
+	cluster *corev1.StorageCluster,
+) map[corev1.ClusterConditionType]*corev1.ClusterCondition {
+	pxConditions := make(map[corev1.ClusterConditionType]*corev1.ClusterCondition)
+	for _, condition := range cluster.Status.Conditions {
+		if condition.Source == constants.PortworxComponentName {
+			pxConditions[condition.Type] = condition.DeepCopy()
+		}
+	}
+	return pxConditions
+}
+
+// getStorageClusterState iterates storage cluster condition list and returns the cluster state
+func getStorageClusterState(
+	cluster *corev1.StorageCluster,
+) corev1.ClusterState {
+	if len(cluster.Status.Conditions) == 0 {
+		return corev1.ClusterStateInit
+	}
+
+	// TODO: put more components into consideration in the future, currently the list only contains PX conditions
+	pxConditions := getPortworxConditions(cluster)
+	if len(pxConditions) == 0 {
+		return corev1.ClusterStateInit
+	}
+
+	// Portworx delete conditions:
+	// InProgress, Completed, Failed
+	deleteCondition := pxConditions[corev1.ClusterConditionTypeDelete]
+	if deleteCondition != nil {
+		return corev1.ClusterStateUninstall
+	}
+
+	// DaemonSet migration conditions:
+	// Pending, InProgress, Completed
+	migrationCondition := pxConditions[corev1.ClusterConditionTypeMigration]
+	if migrationCondition != nil && migrationCondition.Status != corev1.ClusterConditionStatusCompleted {
```

Comment:
What would happen in case of migration failure? Shouldn't we change the phase to Degraded?

Similarly if the install or pre-flight (upgrade when implemented) phase fails, should we change the phase to Degraded?